Automattic/jetpack
# CRM Automations: Bootstrap system

Fixes https://github.com/Automattic/zero-bs-crm/issues/3263

## Proposed changes:

* Add a new `Bootstrap` class that is solely responsible for:
  * Initialise the automation engine
  * Register "core triggers/conditions/actions" and allow 3. parties to hook into it 
* Remove the need to inject `Automation_Engine` into workflow constructs
  * The intent here is to make it a smoother experience programmatically registering workflows.
  * This also means that we had to add new set/get logger methods to a workflow
* Update existing workflow tests to inject the engine via `Automation_Workflow::set_engine`
* Add a new test that ensures that we throw a fatal error if someone tries to execute a workflow without injecting the engine.
* Add a `is_admin` condition around the automation bootstrap logic so we do not bootstrap all of it on front-end page loads.
  * I've added a `@todo` condition since I believe this could be a general issue for CRM.

### Other information:


## Jetpack product discussion



pbhBOz-2ES-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






* Add the below snippet somewhere in your code (e.g. `0-sandbox.php).
* _(Optional) Enable error log and tail it to verify that we do not throw any errors._
* Go to `/wp-admin/admin.php?page=zbs-add-edit&action=edit&zbstype=contact and add a new contact
  * The status should be anything else than `Customer` to verify the workflow worked.
* Verify that the status was set to `Customer` by the workflow.

```php
// Enable the automation system.
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );

// Programmatically define a workflow that will always set a contacts status to "Customer" when they are first created.
add_filter(
	'jpcrm_automation_workflows',
	function ( $workflows ) {
		$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
			[
				'id'           => 'testing',
				'name'         => __( 'New contact: Set status to "Customer"', 'zero-bs-crm' ),
				'description'  => __( 'This automation will change the status of a contact to "Customer" when they are created.', 'zero-bs-crm' ),
				'category'     => 'contact',
				'triggers'     => [
					\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
				],
				'initial_step' => [
					'slug'            => 'contact_status_condition',
					'class_name'      => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::class,
					'attributes'      => [
						'new_status' => 'Customer',
					],
					'next_step_true'  => null,
					'next_step_false' => null,
				],
			]
		);

		$workflows[ $workflow->get_id() ] = $workflow;

		return $workflows;
	}
);
```
